realm/realm-jni/src/io_realm_internal_TableQuery.cpp

```diff
@@ -970,58 +1114,136 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFindAll(
-              return 0;
-          }
-
-          // run the query
-          TableView tableView(query->find_all(S(start), S(end), S(limit)));
-
-          // handover the result
-          std::unique_ptr<SharedGroup::Handover<TableView>> handover = SG(
-                  bgSharedGroupPtr)->export_for_handover(tableView, MutableSourcePayload::Move);
-          return reinterpret_cast<jlong>(handover.release());
+          return findAllWithHandover(env, bgSharedGroupPtr, std::move(query), start, end, limit);
       } CATCH_STD()
       return 0;
   }
 
+
+
+// Should match the values in Java ArgumentsHolder class
+enum query_type {QUERY_TYPE_FIND_ALL = 0, QUERY_TYPE_DISTINCT = 4, QUERY_TYPE_FIND_ALL_SORTED = 1, QUERY_TYPE_FIND_ALL_MULTI_SORTED = 2};
+
+// batch update of async queries
+JNIEXPORT jlongArray JNICALL Java_io_realm_internal_TableQuery_nativeBatchUpdateQueries
+        (JNIEnv *env, jobject, jlong bgSharedGroupPtr, jlong replicationPtr,
+         jlongArray  handover_queries_array /*list of handover queries*/,
+         jobjectArray  query_param_matrix /*type & params of the query to be updated*/,
+         jobjectArray  multi_sorted_indices_matrix,
+         jobjectArray  multi_sorted_order_matrix)
+{
+    TR_ENTER()
+    try {
+        JniLongArray handover_queries_pointer_array(env, handover_queries_array);
+
+        int number_of_queries = env->GetArrayLength(query_param_matrix);
+        jlong exported_handover_tableview_array[number_of_queries];
+
+        // Step1: Position the shared group at the handover query version so we can import all queries
+
+        // read the first query to determine the version we should use
+        SharedGroup::Handover<Query> *handoverQueryPtr = HO(Query, handover_queries_pointer_array[0]);
+        std::unique_ptr<SharedGroup::Handover<Query>> handoverQuery(handoverQueryPtr);
+        // position this shared group at the specified version
+        SG(bgSharedGroupPtr)->begin_read(handoverQuery->version);
+
+        std::unique_ptr<Query> queries[number_of_queries];
+        // import the first query
+        queries[0] = std::move(SG(bgSharedGroupPtr)->import_from_handover(std::move(handoverQuery)));
+
+        // import the rest of the queries
+        for (int i=1; i < number_of_queries; i++) {
+            std::unique_ptr<SharedGroup::Handover<Query>> handoverQuery(HO(Query, handover_queries_pointer_array[i]));
+            queries[i] = std::move(SG(bgSharedGroupPtr)->import_from_handover(std::move(handoverQuery)));
+        }
+
+        // Step2: Bring the queries into the latest shared group version
+        LangBindHelper::advance_read(*SG(bgSharedGroupPtr), *CH(replicationPtr));
+
+        // Step3: Run & export the queries against the latest shared group
+        for (int i = 0; i < number_of_queries; i++) {
+            JniLongArray query_param_array(env, (jlongArray) env->GetObjectArrayElement(query_param_matrix, i));
```

Comment:
`query_param_array` should have at least 1 entry, which is the type of query we need to run in the background, this is set in `QueryUpdateTask#prepareQueriesParameters`